tests/unit/test_reservation.py

```diff
@@ -261,6 +261,54 @@ def test___session_reserved_using_pin_group___get_connection_by_pin___returns_co
         assert connection.session_info == session_info
 
 
+def test___duplicate_pins___get_connections___returns_single_connection(
+    session_management_client: Mock,
+) -> None:
+    with ExitStack() as stack:
+        grpc_session_infos = create_nifake_session_infos(1)
+        grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin1", site=2, channel="3")
+        reservation = MultiSessionReservation(session_management_client, grpc_session_infos)
+        session_info = stack.enter_context(
+            reservation.initialize_session(construct_session, "nifake")
+        )
+
+        connections = reservation.get_connections(
+            fake_driver.Session, pin_or_relay_names=["Pin1", "Pin1"]
+        )
+
+        assert len(connections) == 1
+        assert connections[0].pin_or_relay_name == "Pin1"
+        assert connections[0].site == 2
+        assert connections[0].channel_name == "3"
+        assert connections[0].session_info == session_info
+
+
+def test___session_reserved___get_connections_by_pin_group___returns_connections(
+    session_management_client: Mock,
+) -> None:
+    with ExitStack() as stack:
+        grpc_session_infos = create_nifake_session_infos(1)
+        grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin1", site=2, channel="3")
+        grpc_session_infos[0].channel_mappings.add(pin_or_relay_name="Pin2", site=2, channel="2")
+        group_mappings: Dict[str, Iterable[str]] = {"PinGroup1": ["Pin1", "Pin2"]}
```

Comment:
The fact that you added this type hint suggests that `Dict[str, Iterable[str]]` is wrong. Look more closely at the error message that caused you to add this.

https://mypy-play.net/?mypy=latest&python=3.12&gist=98c21f5d04cf0f7c6be3ee362f5c5462

```
import typing

x = {"a": ["b", "c"]}
reveal_type(x)

def f(x: typing.Dict[str, typing.Iterable[str]]) -> None:
    pass

f(x)
```

```
main.py:4: note: Revealed type is "builtins.dict[builtins.str, builtins.list[builtins.str]]"
main.py:9: error: Argument 1 to "f" has incompatible type "dict[str, list[str]]"; expected "dict[str, Iterable[str]]"  [arg-type]
main.py:9: note: "Dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
main.py:9: note: Consider using "Mapping" instead, which is covariant in the value type
Found 1 error in 1 file (checked 1 source file)
```

Since `dict` is invariant, you can use either:
- `Mapping[str, Iterable[str]]`: works because `Mapping`'s value type is covariant
- `Dict[str, List[str]]`: works because it exactly matches the types we are using

Reply:
BTW, mutable collections are probably invariant because otherwise they would have to be covariant for reads and contravariant for writes. Otherwise, you could upcast `Dict[str, List[str]]` to `Mapping[str, Iterable[str]]` and insert an `Iterable[str]` that isn't a `List[str]`, such as `Tuple[str, ...]`, `str`, or a generator.

Reply:
Delete the type hint from the test and fix the type hint in the BaseReservation class:

```
class BaseReservation(_BaseSessionContainer):
    """Manages session reservation."""
    def __init__(
        self,
        session_management_client: SessionManagementClient,
        session_info: Sequence[session_management_service_pb2.SessionInformation],
        multiplexer_session_info: Optional[
            Sequence[session_management_service_pb2.MultiplexerSessionInformation]
        ] = None,
        pin_or_relay_group_mappings: Optional[Dict[str, Iterable[str]]] = None,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        reserved_pin_or_relay_names: Union[str, Iterable[str], None] = None,
```

Reply:
Thanks for pointing it out. I've now updated.